contractsV5/KyberHintHandler.sol

```diff
@@ -216,84 +247,103 @@ contract KyberHintParser is IKyberHint, Utils {
         bytes memory hint,
         TradeHint memory tradeHint,
         uint indexToContinueFrom,
-        bool isTokenToEth
+        bool isTokenToEth,
+        bool isExternal
     )
         internal
         view
     {
         bytes memory opcode = hint.slice(indexToContinueFrom, 1);
         bytes32 opcodeHash = keccak256(opcode);
-        ReservesHint memory reserves;
+        ReservesHint memory reservesHint;
 
         if (isTokenToEth) {
-            reserves = tradeHint.tokenToEthReserves;
+            reservesHint = tradeHint.tokenToEthReserves;
         } else {
-            reserves = tradeHint.ethToTokenReserves;
+            reservesHint = tradeHint.ethToTokenReserves;
         }
 
         indexToContinueFrom += 1;
+        reservesHint.invalidHint = false;
 
         if (opcodeHash == keccak256(END_OPCODE)) {
             return;
         } else if (opcodeHash == keccak256(SEPARATOR)) {
-            decodeOperation(hint, tradeHint, indexToContinueFrom, false);
-        } else if (opcodeHash == keccak256(parseOpcode(TradeType.MaskIn))) {
-            reserves.tradeType = TradeType.MaskIn;
-            (reserves.addresses, reserves.splitValuesBps, indexToContinueFrom) = decodeReservesFromHint(false, hint, indexToContinueFrom);
-            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
-        } else if (opcodeHash == keccak256(parseOpcode(TradeType.MaskOut))) {
-            reserves.tradeType = TradeType.MaskOut;
-            (reserves.addresses, reserves.splitValuesBps, indexToContinueFrom) = decodeReservesFromHint(false, hint, indexToContinueFrom);
-            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
-        } else if (opcodeHash == keccak256(parseOpcode(TradeType.Split))) {
-            reserves.tradeType = TradeType.Split;
-            (reserves.addresses, reserves.splitValuesBps, indexToContinueFrom) = decodeReservesFromHint(true, hint, indexToContinueFrom);
-            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
-        } else {
-            revert("Invalid hint opcode");
+            decodeOperation(hint, tradeHint, indexToContinueFrom, false, isExternal);
+        } else if (opcodeHash == keccak256(encodeOpcode(TradeType.MaskIn))) {
```

Comment:
does this method have better gas consumption then string compare?
we can consider having value of keccak256(encodeOpcode(TradeType.MaskIn)) ready as constant. and save run time.
unless compiler is smart enough to do it...
please check.

Reply:
Using the keccak(opcode) constants instead for comparison.